File: src/L03/Fluffy.hs
module L03.Fluffy where

import L01.Id
import L01.Optional
import L01.Validation
import L02.List

class Fluffy f where
  furry :: (a -> b) -> f a -> f b

-- Exercise 1
-- Relative Difficulty: 1
instance Fluffy Id where
  furry =
    mapId

-- Exercise 2
-- Relative Difficulty: 2
instance Fluffy List where
  furry =
    maap

-- Exercise 3
-- Relative Difficulty: 2
instance Fluffy Optional where
  furry =
    mapOptional

-- Exercise 4
-- Relative Difficulty: 3
instance Fluffy ((->) t) where
  furry f g =
    \x -> f (g x)

-----------------------
-- SUPPORT LIBRARIES --
-----------------------

instance Fluffy [] where
  furry = fmap

File: src/L03/Fluffy.scala
package L03

import L01._
import L02._

trait Fluffy[F[_]] {
  def furry[A, B](f: A => B): F[A] => F[B]

  // Provided as a synonym for `furry` but with the arguments flipped.
  // This helps the type-inferencer.
  final def apply[A, B](a: F[A])(f: A => B): F[B] =
    furry(f)(a)
}

object Fluffy {
  // Exercise 1
  // Relative Difficulty: 1
  implicit val IdFluffy: Fluffy[Id] =
    sys.error("todo")

  // Exercise 2
  // Relative Difficulty: 2
  implicit val ListFluffy: Fluffy[List] =
    sys.error("todo")

  // Exercise 3
  // Relative Difficulty: 2
  implicit val OptionalFluffy: Fluffy[Optional] =
    sys.error("todo")

  // Exercise 4
  // Relative Difficulty: 3
  implicit def Function1Fluffy[T]: Fluffy[({type l[a] = T => a})#l] =
    sys.error("todo")
  
  ///////////////////////
  // SUPPORT LIBRARIES //
  ///////////////////////

  implicit val ScalaListFluffy: Fluffy[scala.List] =
    new Fluffy[scala.List] {
      def furry[A, B](f: A => B) =
        _ map f
    }

}
